sumpy/tools.py

```diff
@@ -976,19 +981,198 @@ def wait(self):
+                assignee=x[(*broadcast_dims, i3)],
+                expression=x[(*broadcast_dims, i3)] / n,
+                depends_on=frozenset([f"update_{len(factors) - 1}"]),
+            ),
+        ]
+
+    if name is None:
+        if inverse:
+            name = f"ifft_{n}"
+        else:
+            name = f"fft_{n}"
+
+    knl = lp.make_kernel(domains, insns,
+        kernel_data=[
+            lp.GlobalArg("x", shape=shape, is_input=False, is_output=True,
+                dtype=complex_dtype),
+            lp.GlobalArg("y", shape=shape, is_input=True, is_output=False,
+                dtype=complex_dtype),
+            lp.TemporaryVariable("exp_table", shape=(n,),
+                dtype=complex_dtype),
+            lp.TemporaryVariable("temp", shape=(n,),
+                dtype=complex_dtype),
+            ...],
+        name=name,
+        fixed_parameters=fixed_parameters,
+        lang_version=lp.MOST_RECENT_LANGUAGE_VERSION,
+        index_dtype=index_dtype,
+    )
+    if broadcast_dims:
+        knl = lp.split_iname(knl, "j0", 32, inner_tag="l.0", outer_tag="g.0")
+        knl = lp.add_inames_for_unused_hw_axes(knl)
+
+    return knl
+
+
+def _get_fft_backend(queue):
+    env_val = os.environ.get("SUMPY_FFT_BACKEND", None)
+    if env_val:
+        if env_val not in ["loopy", "vkfft"]:
+            raise ValueError("Expected 'loopy' or 'vkfft' for SUMPY_FFT_BACKEND. "
+                   f"Found {env_val}.")
+        return env_val
+
+    try:
+        import pyvkfft.opencl  # noqa: F401
+    except ImportError:
+        warnings.warn("VkFFT not found. FFT runs will be slower.")
+        return "loopy"
+
     if queue.properties & cl.command_queue_properties.OUT_OF_ORDER_EXEC_MODE_ENABLE:
-        raise RuntimeError("VkFFT does not support out of order queues yet.")
+        warnings.warn("VkFFT does not support out of order queues yet. "
+            "Falling back to slower implementation.")
+        return "loopy"
+
+    if (sys.platform == "darwin"
+            and queue.context.devices[0].platform.name
+            == "Portable Computing Language"):
+        warnings.warn("Pocl miscompiles some VkFFT kernels. "
+            "Falling back to slower implementation.")
+        return "loopy"
+
+    return "vkfft"
 
+
+def get_opencl_fft_app(queue, shape, dtype, inverse):
+    """Setup an object for out-of-place FFT on with given shape and dtype
+    on given queue.
+    """
     assert dtype.type in (np.float32, np.float64, np.complex64,
                            np.complex128)
 
-    from pyvkfft.opencl import VkFFTApp
-    app = VkFFTApp(shape=shape, dtype=dtype, queue=queue, ndim=1, inplace=False)
-    return app
+    backend = _get_fft_backend(queue)
+
+    if backend == "loopy":
+        return loopy_fft(shape, inverse=inverse, complex_dtype=dtype.type)
+    else:
+        from pyvkfft.opencl import VkFFTApp
+        app = VkFFTApp(shape=shape, dtype=dtype, queue=queue, ndim=1, inplace=False)
```

Comment:
How come this doesn't look at `inverse`?

Reply:
VkFFT generated code for both inverse and non-inverse at the same time.